packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,345 @@
+    },
+  });
+}
+
+/** @type { Harness | null } */
+let theHarness = null; // ISSUE: ambient
+
+/**
+ * @param {(msg: TapMessage) => void} send
+ *
+ * @typedef { ReturnType<typeof createHarness>} Harness
+ */
+function createHarness(send) {
+  let testNum = 0;
+  let passCount = 0;
+  /** @type {((ot: { context: Object }) => Promise<void>)[]} */
+  let beforeHooks = [];
+  /** @type {(() => Promise<void>)[]} */
+  let suitesToRun = [];
+  const context = {};
+
+  /**
+   * @returns { Summary }
+   * @typedef {import('./avaXS').Summary} Summary
+   */
+  function summary() {
+    return {
+      pass: passCount,
+      fail: testNum - passCount,
+      total: testNum,
+    };
+  }
+
+  const it = freeze({
+    send,
+    get context() {
+      return context;
+    },
+    /** @type {(l: string, f: () => Promise<void>) => void } */
+    before(_label, hook) {
+      beforeHooks.push(hook);
+    },
+    /** @type { (ok: boolean) => number } */
+    finish(ok) {
+      testNum += 1;
+      if (ok) {
+        passCount += 1;
+      }
+      return testNum;
+    },
+    /** @type { (f: () => Promise<void>) => Promise<void> } */
+    async defer(thunk) {
+      suitesToRun.push(thunk);
+    },
+    summary,
+    async result() {
+      for await (const hook of beforeHooks) {
+        await hook({ context });
+      }
+      beforeHooks = [];
+      for await (const suite of suitesToRun) {
+        await suite();
+      }
+      suitesToRun = [];
+
+      return summary();
+    },
+  });
+
+  if (!theHarness) {
+    theHarness = it;
+  }
+  return it;
+}
+
+/**
+ * @param {*} exc
+ * @param {Expectation} expectation
+ *
+ * @typedef {{ instanceOf: Function } | { message: string | RegExp }=} Expectation
+ */
+function checkExpectation(exc, expectation) {
+  if (!expectation) return true;
+  if ('instanceOf' in expectation) return exc instanceof expectation.instanceOf;
+  if ('message' in expectation) {
+    const { message } = expectation;
+    return typeof message === 'string'
+      ? exc.message === message
+      : exc.message.match(message);
+  }
+  throw Error(`not implemented: ${JSON.stringify(expectation)}`);
+}
+
+/**
+ * @param {Harness} htest
+ * @param {TapFormat} out
+ *
+ * @typedef {ReturnType<typeof makeTester>} Tester
+ */
+function makeTester(htest, out) {
+  /** @type {number?} */
+  let pending;
+
+  /** @type {(r: boolean, info?: string) => void} */
+  function assert(result, info) {
+    if (typeof pending === 'number') {
+      pending -= 1;
+    }
+    const testNum = htest.finish(result);
+    if (result) {
+      out.ok(testNum, info);
+    } else {
+      out.notOk(testNum, info);
+    }
+  }
+
+  function truthy(/** @type {unknown} */ value, msg = 'should be truthy') {
+    assert(!!value, msg);
+  }
+
+  /** @type {(a: unknown, e: unknown) => void } */
+  function deepEqTest(actual, expected) {
+    try {
+      assert(deepEqual(actual, expected), 'should be deep equal');
+    } catch (detail) {
```

Comment:
Don't call an error object `details`. That is totally confusing given the way the `assert` package uses `details`.

Reply:
I was there first! :)

Will fix. Suggestions?

Reply:
Best is `reason`. Convention comes from promises. It is what a rejected promise is rejected with. Better than `err` or `error` because it isn't necessarily an error, although it typically is. Same with the thrown/caught thing.